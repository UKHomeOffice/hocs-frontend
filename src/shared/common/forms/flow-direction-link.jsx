import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ApplicationConsumer } from '../../contexts/application.jsx';
import { Link } from 'react-router-dom';
import axios from 'axios';
import { withRouter } from 'react-router-dom';
import {
    setError,
    updateApiStatus,
    clearApiStatus,
} from '../../contexts/actions/index.jsx';
import status from '../../helpers/api-status.js';

export class FlowDirectionLink extends Component {

    handleClick(e) {
        e.preventDefault();
        const { dispatch, history } = this.props;
        const endpoint = `/api/form/case/${this.props.caseId}/stage/${this.props.stageId}/direction/${this.props.flowDirection}`;
        return dispatch(updateApiStatus(status.FLOW_DIRECTION_LINK_REQUEST))
            .then(() => {
                axios.get(endpoint)
                    .then(response => {
                        if (response.data.errors) {
                            dispatch(updateApiStatus(status.FLOW_DIRECTION_LINK_FAILURE));
                        } else {
                            dispatch(updateApiStatus(status.FLOW_DIRECTION_LINK_SUCCESS));
                            history.push(response.data.redirect);
                        }
                    }).then(() => dispatch(clearApiStatus()));
            })
            .catch(error => {
                dispatch(updateApiStatus(status.FLOW_DIRECTION_LINK_FAILURE))
                    .then(() => dispatch(setError(error.response)));
            });
    }

    render() {
        const {
            action,
            className,
            disabled,
            label
        } = this.props;
        return (
            <p>
                <Link
                    className={`govuk-body govuk-link${className ? ' ' + className : ''}`}
                    disabled={disabled}
                    onClick={e => this.handleClick(e)}
                    to={action}
                >{label}</Link>
            </p>
        );
    }
}

FlowDirectionLink.propTypes = {
    caseId: PropTypes.string.isRequired,
    stageId: PropTypes.string.isRequired,
    action: PropTypes.string.isRequired,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    dispatch: PropTypes.func.isRequired,
    label: PropTypes.string,
    history: PropTypes.object.isRequired,
    flowDirection: PropTypes.string.isRequired,
};

FlowDirectionLink.defaultProps = {
    disabled: false,
    label: 'Update'
};

const WrappedButton = props => (
    <ApplicationConsumer>
        {({ dispatch }) => <FlowDirectionLink {...props} dispatch={dispatch} />}
    </ApplicationConsumer>
);

const FlowDirectionLinkWithRouter = withRouter(WrappedButton);

export default FlowDirectionLinkWithRouter;
