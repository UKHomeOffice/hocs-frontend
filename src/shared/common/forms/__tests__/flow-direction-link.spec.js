import React from 'react';
import { FlowDirectionLink } from '../flow-direction-link.jsx';
import { MemoryRouter } from 'react-router-dom';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';

describe('Form Flow Direction Link component', () => {

    const mockDispatch = jest.fn(() => Promise.resolve());
    const mockHistory = {};

    beforeEach(() => {
        mockDispatch.mockReset();
    });

    it('should render with default props', () => {
        const wrappedButton = render(
            <FlowDirectionLink action="/SOME/URL" dispatch={mockDispatch}
                history={mockHistory} caseId='caseId123' stageId='stageId123' flowDirection='back' />,
            { wrapper: MemoryRouter }
        );
        expect(wrappedButton).toBeDefined();
    });

    it('should render disabled when isDisabled is passed', () => {
        render(
            <FlowDirectionLink action="/SOME/URL" dispatch={mockDispatch}
                history={mockHistory} caseId='caseId123' stageId='stageId123' flowDirection='back' disabled={true} />,
            { wrapper: MemoryRouter }
        );
        expect(screen.getByRole('link')).toHaveAttribute('disabled');
    });

    it('should render with correct label when passed', () => {
        render(
            <FlowDirectionLink label={'My Button'} action="/SOME/URL" dispatch={mockDispatch}
                history={mockHistory} caseId='caseId123' stageId='stageId123' flowDirection='back' />,
            { wrapper: MemoryRouter }
        );
        expect(screen.getByText('My Button')).toBeInTheDocument();
    });
});
