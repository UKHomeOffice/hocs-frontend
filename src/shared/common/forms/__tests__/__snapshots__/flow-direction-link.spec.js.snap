// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Flow Direction Link component should render disabled when isDisabled is passed 1`] = `
<FlowDirectionLink
  action="/SOME/URL"
  caseId="caseId123"
  disabled={true}
  dispatch={[MockFunction]}
  flowDirection="back"
  history={Object {}}
  label="Update"
  stageId="stageId123"
>
  <p>
    <Link
      className="govuk-body govuk-link"
      disabled={true}
      onClick={[Function]}
      to="/SOME/URL"
    >
      <LinkAnchor
        className="govuk-body govuk-link"
        disabled={true}
        href="/SOME/URL"
        navigate={[Function]}
        onClick={[Function]}
      >
        <a
          className="govuk-body govuk-link"
          disabled={true}
          href="/SOME/URL"
          onClick={[Function]}
        >
          Update
        </a>
      </LinkAnchor>
    </Link>
  </p>
</FlowDirectionLink>
`;

exports[`Form Flow Direction Link component should render with additional styles when className is passed 1`] = `
<FlowDirectionLink
  action="/SOME/URL"
  caseId="caseId123"
  className="test-class"
  disabled={false}
  dispatch={[MockFunction]}
  flowDirection="back"
  history={Object {}}
  label="Update"
  stageId="stageId123"
>
  <p>
    <Link
      className="govuk-body govuk-link test-class"
      disabled={false}
      onClick={[Function]}
      to="/SOME/URL"
    >
      <LinkAnchor
        className="govuk-body govuk-link test-class"
        disabled={false}
        href="/SOME/URL"
        navigate={[Function]}
        onClick={[Function]}
      >
        <a
          className="govuk-body govuk-link test-class"
          disabled={false}
          href="/SOME/URL"
          onClick={[Function]}
        >
          Update
        </a>
      </LinkAnchor>
    </Link>
  </p>
</FlowDirectionLink>
`;

exports[`Form Flow Direction Link component should render with correct label when passed 1`] = `null`;

exports[`Form Flow Direction Link component should render with default props 1`] = `
<FlowDirectionLink
  action="/SOME/URL"
  caseId="caseId123"
  disabled={false}
  dispatch={[MockFunction]}
  flowDirection="back"
  history={Object {}}
  label="Update"
  stageId="stageId123"
>
  <p>
    <Link
      className="govuk-body govuk-link"
      disabled={false}
      onClick={[Function]}
      to="/SOME/URL"
    >
      <LinkAnchor
        className="govuk-body govuk-link"
        disabled={false}
        href="/SOME/URL"
        navigate={[Function]}
        onClick={[Function]}
      >
        <a
          className="govuk-body govuk-link"
          disabled={false}
          href="/SOME/URL"
          onClick={[Function]}
        >
          Update
        </a>
      </LinkAnchor>
    </Link>
  </p>
</FlowDirectionLink>
`;
